contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,340 @@
+        }
+
+        // Calculate amount of pending veJOE based on:
+        // 1. Seconds elapsed since last reward timestamp
+        // 2. Generation rate that the user is receiving
+        // 3. Current amount of user's staked JOE
+        uint256 pendingVeJoe;
+
+        if (block.timestamp <= user.boostEndTimestamp) {
+            // If the current timestamp is less than or equal to the user's `boostEndTimestamp`,
+            // that means the user is currently receiving boosted benefits so they should receive
+            // `boostedGenerationRate`.
+            uint256 accVeJoePerJoe = secondsElapsed.mul(boostedGenerationRate).div(PRECISION);
+            pendingVeJoe = accVeJoePerJoe.mul(user.balance);
+        } else {
+            if (user.boostEndTimestamp != 0) {
+                // If `user.boostEndTimestamp != 0` then, we know for certain that
+                // `user.boostEndTimestamp >= user.lastRewardTimestamp`.
+                // Proof by contradiction:
+                // 1. Assume that `user.boostEndTimestamp != 0` and
+                //    `user.boostEndTimestamp < user.lastRewardTimestamp`.
+                // 2. That means that at time `user.lastRewardTimestamp`, the user claimed
+                //    some veJOE. Furthermore, we know that anytime a user claims some veJOE,
+                //    if the current timestamp is greater than or equal to `user.boostEndTimestamp`,
+                //    we will update `user.boostEndTimestamp` to be `0` (see `_claim` method).
+                // 3. This means that `user.boostEndTimestamp` should be `0` but that contradicts our
+                //    assumption that `user.boostEndTimestamp != 0`
+                // QED.
+                // With this, we now know `0 < user.lastRewardTimestamp <= user.boostEndTimestamp < block.timestamp`,
+                // which will allow us to perform the following logic safely.
+
+                // If the `block.timestamp > user.boostEndTimestamp` and `boostEndTimestamp != 0`,
+                // that means the user's boosted benefits ended sometime between their `lastRewardTimestamp`
+                // and now.
+                // In this case, we need to properly provide them the boosted generation rate for
+                // those `boostEndTimestamp - lastRewardTimestamp` seconds.
+                uint256 boostedTimeElapsed = user.boostEndTimestamp.sub(user.lastRewardTimestamp);
+                uint256 boostedAccVeJoePerJoe = boostedTimeElapsed.mul(boostedGenerationRate).div(PRECISION);
+                uint256 boostedPendingVeJoe = boostedAccVeJoePerJoe.mul(user.balance);
+
+                uint256 baseTimeElapsed = block.timestamp.sub(user.boostEndTimestamp);
+                uint256 baseAccVeJoePerVeJoe = baseTimeElapsed.mul(baseGenerationRate).div(PRECISION);
+                uint256 basePendingVeJoe = baseAccVeJoePerVeJoe.mul(user.balance);
+
+                pendingVeJoe = boostedPendingVeJoe.add(basePendingVeJoe);
```

Comment:
pendingVeJoe = boostedPendingVeJoe.add(basePendingVeJoe).div(PRECISION)

Now we divide it because those 2 numbers are now in 1e36 (18 + `veJOE.decimals()`), and we need them to have `veJOE.decimals()` decimals.

It may be easier to understand if we divide at L289 and L285, but it would cost twice more gas than doing it here